io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -947,6 +952,26 @@ impl LvsLvol for Lvol {
             Some(self.name()),
             Some(self.name()),
             Some(Utc::now().to_string()),
+            false,
         )
     }
+    /// Wrapper function to destroy replica and its associated snapshot if
+    /// replica is identified as last clone.
+    async fn destroy_replica(mut self) -> Result<String, Error> {
+        let snapshot_lvol = self.is_snapshot_clone();
+        let name = self.name();
+        self.destroy().await?;
+
+        // If destroy replica is a snapshot clone and it is the last
+        // clone from the snapshot, destroy the snapshot
+        // if it is already marked as discardedSnapshot.
+        if let Some(snapshot_lvol) = snapshot_lvol {
```

Comment:
Let's say we crash after self.destroy and before we destroy snapshot_vol, then we "leak" it?
Should we also do this on pool import, so after the crash we "garbage collect"?
And can we also add a test to check this works?

Reply:
Yes, There will be leak if just clone.destroy() is success and there is a crash after that before snapshot destroy. Whether control plane will retry the replica destroy gRPC again after pool import ? 
If control plane trigger the replica destroy gRPC again, then "garbage collection" can be handled in replica destroy flow.

Reply:
Ah if destroy_replica call has enough information for us to do it then, I guess that could work too!
Not sure which way would be better, any thoughts @mtzaurus ?

Reply:
In destroy_replica request, We will get the replica uuid. If replica_uuid not found in the system, then rather failing the request, We can iterate through bdev to get lvol and match with `CloneXAttr.SourceUuid` with input replica_uuid to destroy it.

Reply:
Currently handled it as part of the destroy replica gRPC flow, when the retry of the destroy replica happens, a silent function (Without catching the snapshot destroy error) added to clean the discarded snapshot. The other way to send back the error code if destroy snapshot fails and CP will retry again. Not sure if We need to consider it @mtzaurus  @tiagolobocastro please suggest.

Reply:
> Let's say we crash after self.destroy and before we destroy snapshot_vol, then we "leak" it? Should we also do this on pool import, so after the crash we "garbage collect"? And can we also add a test to check this works?

We need to perform "garbage collection" also upon I/O engine start, to handle such a situation properly.

Reply:
Yeah I also think it's better to do this on import rather than per-request later on.